test/PredictiveDeployer.t.sol

```diff
@@ -1,90 +1,155 @@
+        vm.startPrank(sender);
+        address expectedChild = IPredictiveDeployer(address(proxy)).getAddress(wallet.addr, childBytecode);
+        vm.expectEmit(true, true, true, true, address(proxy));
+        emit Deploy(wallet.addr, expectedChild, keccak256(childBytecode), currentNonce);
 
-    //     // Expectations
-    //     vm.startPrank(sender);
-    //     address expectedChild = predictive_deployer.getAddress(messageHash, signature, childBytecode);
-    //     vm.expectEmit(true, true, true, true, address(predictive_deployer));
-    //     emit Deploy(sender, expectedChild, keccak256(childBytecode), currentNonce);
+        // Act
+        address actualChild = IPredictiveDeployer(address(proxy)).deploy(wallet.addr, signature, childBytecode);
+        vm.stopPrank();
 
-    //     // Act
-    //     address actualChild = predictive_deployer.deploy(messageHash, signature, childBytecode);
-    //     vm.stopPrank();
+        // Assertions
+        assertEq(actualChild, expectedChild);
+    }
 
-    //     // Assertions
-    //     assertEq(actualChild, expectedChild);
-    // }
+    function test_CannotDeployReplay(uint256 pkNum) public {
+        // Setup
+        VmSafe.Wallet memory wallet = vm.createWallet(uint256(keccak256(abi.encodePacked(uint256(pkNum)))));
+
+        uint256 currentNonce = IPredictiveDeployer(address(proxy)).userNonces(wallet.addr);
+
+        // Get signature information
+        bytes32 txHash = IPredictiveDeployer(address(proxy)).getTransactionHash(wallet.addr, currentNonce);
+
+        bytes32 messageHash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", txHash));
+
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(wallet.privateKey, messageHash);
+        bytes memory signature = abi.encodePacked(r, s, v);
+
+        // Deploy once
+        IPredictiveDeployer(address(proxy)).deploy(wallet.addr, signature, childBytecode);
+
+        // Act: attempt replay
+        vm.expectRevert(PredictiveDeployer.Unauthorized.selector);
+        IPredictiveDeployer(address(proxy)).deploy(wallet.addr, signature, childBytecode);
+    }
+
+    function test_CannotDeployWithoutApproval(uint256 pkNum, address invalidPrincipal) public {
+        // Setup
+        VmSafe.Wallet memory wallet = vm.createWallet(uint256(keccak256(abi.encodePacked(uint256(pkNum)))));
+
+        // Failing condition: principal is not the signer
+        vm.assume(invalidPrincipal != wallet.addr);
+
+        uint256 currentNonce = IPredictiveDeployer(address(proxy)).userNonces(wallet.addr);
+
+        // Get signature information
+        bytes32 txHash = IPredictiveDeployer(address(proxy)).getTransactionHash(wallet.addr, currentNonce);
+
+        bytes32 messageHash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", txHash));
+
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(wallet.privateKey, messageHash);
+        bytes memory signature = abi.encodePacked(r, s, v);
+
+        // Act: attempt replay
```

Comment:
Changed this to `// Act: attempt with invalid principal locally`. Waiting to push until you finish your reviews @rahul0eth @0xernesto